packages/modules/viewer-react/src/utilities/MakeCancellable.ts

```diff
@@ -0,0 +1,57 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+export function makeCancellable(
+  fn: (args?: any[]) => Generator<Promise<any> | Promise<void>, void, any>,
+  ...args: any[]
+): { promise: Promise<any>; cancel: () => void } {
+  const gen = fn(...args);
+  let cancelled = false;
+  let cancel: () => void = () => {
+    // nothing to cancel
+  };
+  const promise = new Promise((resolve, reject) => {
+    cancel = () => {
+      cancelled = true;
+      reject({ reason: "cancelled" });
+    };
+
+    onFulfilled();
+
+    function onFulfilled(res?: any) {
+      if (!cancelled) {
+        let result;
+        try {
+          result = gen.next(res);
+        } catch (e) {
+          return reject(e);
+        }
+        next(result);
+        return null;
+      }
+    }
+
+    function onRejected(err: any) {
+      let result;
+      try {
+        result = gen.throw(err);
```

Comment:
is this necessary? can't just reject and move on?

Reply:
rejecting the promise doesn't necessarily stop the generator from further executing if it is accessed again. The reject will only reject the current promise we're wrapped in, but doesn't prevent the generator function from sticking around in a suspended state, which could also cause a memory leak